File: packages/dukan_core/lib/domain/purchasing.dart
/// Purchasing domain. Mirrors docs/domain/purchasing.md and
/// server/dukan/domain/purchasing.py. Phase 4 uses a direct goods receipt
/// (no full PO lifecycle): receiving increments stock at cost and bills the
/// supplier. The supplier balance is derived from an append-only ledger.
library;

import '../shared/errors.dart';
import 'numbers.dart' show moneyMax;
import 'sales.dart' show lineTotalFits, lineTotalMinor;

enum SupplierEntryType { bill, payment, adjustment }

final class Supplier {
  const Supplier({
    required this.id,
    required this.name,
    this.phone,
    this.currency = 'AFN',
    this.isActive = true,
    this.version = 0,
  });

  final String id;
  final String name;
  final String? phone;
  final String currency;
  final bool isActive;
  final int version;
}

final class SupplierLedgerEntry {
  const SupplierLedgerEntry({
    required this.id,
    required this.supplierId,
    required this.type,
    required this.amountMinor,
    required this.currency,
    required this.occurredAt,
  });

  final String id;
  final String supplierId;
  final SupplierEntryType type;
  final int amountMinor;
  final String currency;
  final DateTime occurredAt;

  /// Bills add (shop owes), payments subtract.
  int get signed => switch (type) {
        SupplierEntryType.bill => amountMinor,
        SupplierEntryType.adjustment => amountMinor,
        SupplierEntryType.payment => -amountMinor,
      };
}

int supplierBalance(Iterable<SupplierLedgerEntry> entries) =>
    entries.fold(0, (sum, e) => sum + e.signed);

/// One line of a goods receipt: [qtyMinor] in the unit's minor granularity
/// (10^[decimalPlaces]) at [unitCostMinor] per whole unit.
final class ReceiptLine {
  const ReceiptLine({
    required this.productId,
    required this.qtyMinor,
    required this.unitCostMinor,
    required this.decimalPlaces,
  });
  final String productId;
  final int qtyMinor;
  final int unitCostMinor;
  final int decimalPlaces;

  /// Quantity × cost at the unit's scale, ROUND_HALF_UP like a sale line: 2.500
  /// kg at 40.00 is 100.00.
  int get lineCost => lineTotalMinor(unitCostMinor, qtyMinor, decimalPlaces);
}

/// A received line adds a positive quantity at a cost of zero or more (zero: a
/// quantity-only receipt), costing at most [moneyMax]. Raises [ValidationError]
/// `GRN_LINE_INVALID` or `GRN_TOTAL_TOO_LARGE`.
void assertReceivable(ReceiptLine line) {
  if (line.qtyMinor <= 0 || line.unitCostMinor < 0) {
    throw ValidationError('GRN_LINE_INVALID', {
      'product_id': line.productId, 'qty': line.qtyMinor, 'cost': line.unitCostMinor,
    });
  }
  if (!lineTotalFits(line.unitCostMinor, line.qtyMinor, line.decimalPlaces)) {
    throw ValidationError('GRN_TOTAL_TOO_LARGE', {'product_id': line.productId});
  }
}

/// What a receipt of valid lines bills: their costs, at most [moneyMax].
/// Raises [ValidationError] `GRN_TOTAL_TOO_LARGE`.
int receiptTotal(Iterable<ReceiptLine> lines) {
  var total = 0;
  for (final l in lines) {
    total += l.lineCost; // each at most moneyMax, so the sum cannot wrap first
    if (total > moneyMax) throw ValidationError('GRN_TOTAL_TOO_LARGE', {});
  }
  return total;
}

/// A payment to a supplier is a positive amount, no more than the shop owes
/// them. Raises [ValidationError] `SUPPLIER_PAYMENT_INVALID` or [ConflictError]
/// `SUPPLIER_OVERPAYMENT`.
void assertSupplierPaymentValid({required int amountMinor, required int balanceMinor}) {
  if (amountMinor <= 0) throw ValidationError('SUPPLIER_PAYMENT_INVALID', {'amount': amountMinor});
  if (amountMinor > balanceMinor) {
    throw ConflictError('SUPPLIER_OVERPAYMENT', {'amount': amountMinor, 'balance': balanceMinor});
  }
}

File: server/dukan/domain/purchasing.py
"""Purchasing domain. Mirrors packages/dukan_core/lib/domain/purchasing.dart."""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass
from datetime import datetime
from enum import StrEnum

from dukan.domain.sales import line_total_minor
from dukan.shared.errors import ValidationError


class SupplierEntryType(StrEnum):
    BILL = "bill"
    PAYMENT = "payment"
    ADJUSTMENT = "adjustment"


@dataclass(frozen=True, slots=True)
class Supplier:
    id: str
    name: str
    phone: str | None = None
    currency: str = "AFN"
    is_active: bool = True
    version: int = 1


@dataclass(frozen=True, slots=True)
class SupplierLedgerEntry:
    id: str
    supplier_id: str
    type: SupplierEntryType
    amount_minor: int
    currency: str
    occurred_at: datetime

    @property
    def signed(self) -> int:
        return -self.amount_minor if self.type is SupplierEntryType.PAYMENT else self.amount_minor


def supplier_balance(entries: Iterable[SupplierLedgerEntry]) -> int:
    return sum(e.signed for e in entries)


@dataclass(frozen=True, slots=True)
class ReceiptLine:
    """qty_minor in the unit's minor granularity (10**decimal_places) at
    unit_cost_minor per whole unit."""

    product_id: str
    qty_minor: int
    unit_cost_minor: int
    decimal_places: int

    @property
    def line_cost(self) -> int:
        """Quantity x cost at the unit's scale, ROUND_HALF_UP like a sale line:
        2.500 kg at 40.00 is 100.00."""
        return line_total_minor(self.unit_cost_minor, self.qty_minor, self.decimal_places)


def assert_receivable(line: ReceiptLine) -> None:
    """A received line adds a positive quantity at a cost of zero or more (zero:
    a quantity-only receipt). Raises ValidationError GRN_LINE_INVALID."""
    if line.qty_minor <= 0 or line.unit_cost_minor < 0:
        raise ValidationError(
            "GRN_LINE_INVALID", product_id=line.product_id, qty=line.qty_minor,
            cost=line.unit_cost_minor,
        )
